File: BrianLee/Week5/2022-10-13-Decode-Ways.java
class Solution {
    public int numDecodings(String s) {
        return backtracking(s, 0, new int[s.length()]);

    }

    private int backtracking(String s, int cur, int[] mem) {
        if(cur == s.length()) return 1;
        if(mem[cur] != 0) return mem[cur];

        char c = s.charAt(cur);
        int count = 0;
        if(c != '0') {
            count = backtracking(s, cur+1, mem);
            if(cur+1 < s.length() && Integer.parseInt(c+""+s.charAt(cur+1)) <= 26) {
                count += backtracking(s, cur+2, mem);
            }
        } else {
            return 0;
        }
        return mem[cur] = count;
    }
}

File: Chris/2022-10-13-Decode Ways.py
class Solution:
    def numDecodings(self, s: str) -> int:
        if s[0] == '0':
            return 0
        
        dp = [0] * (len(s)+1)
        dp[len(s)-1] = 1 if s[-1] != '0' else 0
        
        dp[len(s)] = 1
        
        for i in range(len(s)-2,-1,-1):
            
            if s[i] == '1':

                dp[i] = dp[i+1] + dp[i+2]
                
                
            elif s[i] == '2':
                if s[i+1] not in ['7','8','9']:
                    dp[i] = dp[i+1] + dp[i+2]
                else:
                    dp[i] = dp[i+1]
            
            else:
                if s[i] == '0':
                    dp[i] == 0
                else:
                    dp[i] = dp[i+1]
                    
        return dp[0]
                
                
                    
                
            